### ISSUE TITLE
Work in progress

### ISSUE THREAD

**Issue by owulveryck** [owulveryck/onnx-go#1]
# About

This is just the interface between the ONNX structures/files and the Go ecosystem.
There is also an ongoing effort to implement a backend into [Gorgonia](https://github.com/gorgonia/gorgonia) which is a Computation Lib for Go.

So far, the TensorProto import is partially implemented into the [tensor lib](https://github.com/gorgonia/tensor) but I am waiting for more progress before I ask for a PR and a merge to the master branch.

Regarding the Model and the Graph structures, I have started a POC which is quick'n'dirty by now. (if you are interested, the code is [here](https://github.com/owulveryck/gorgonnx)). My goal is to be able to run the MNIST example. So far I have generated an ExprGraph and I can run it, but the result is wrong. I am doing some bug hunting.

# Next

Once the POC is working, I will do some PR, and start a complete integration process of ONNX into Gorgonia (it may need some tooling and enhanced testing).
Meanwhile, if you have any idea for enhancing the onnx-go repo, please feel free to open an issue or a PR.

cc @jspisak @prasanthpul @lupesko @bddppq

**Comment by thinkski** [comment]
Very cool, hope you continue with this. Being able to convert a model from most any framework into ONNX, then compile (or cross-compile to a different `GOOS` and `GOARCH`) a pure Go executable to perform inference with the model, would make deploying models significantly easier, even if inference was not optimized for speed.

**Comment by owulveryck** [comment]
Thank you very much for this message.

Yes, the project is still actively developed, but as I do it on my free time, it goes not as fast as expected the [gorgonnx](https://github.com/owulveryck/gorgonnx) project was the most advanced part but was suffering from a bug I did not find.

> Debugging is twice as hard as writing the code in the first place. Therefore, if you write the code as cleverly as possible, you are, by definition, not smart enough to debug it. - Kernighan

From the last two weeks I have decided to change the API; the protobuf conversion is not directly exposed; instead I am exposing a ˋunmarshal` function that takes a pointer to gonum’s `GraphBuilder` interface (see the branch ˋgraph-builder` branch)
So far it’s easy to reconstruct the graph (and draw a nice representation with graphviz for example).
It’s also easy to write a backend that will numerically compute the graph.

By now, I am heavily modifying the Gorgonia library to make it compatible with this interface; Gorgonia can do a stable evaluation of the graph *and* automatic differentiation; it could be able to use it for transfer learning).
On top of that, Gorgonia has Cuda binding; the perf on the CPU is not bad, the perf on the GPU is better ;)

**Comment by owulveryck** [comment]
**Tl;DR: onnx-go is working, but the API will change!** 

* The bug with the Gorgonia backend has been fixed (hidden in the `im2col` operator [Issue 268 of Gorgonia](https://github.com/gorgonia/gorgonia/issues/268#issue-413225642)). 
I have modified the test branch `directed-graph` (which have a vendorer version of GorgoniaThe tweaked version of Gorgonia). This branch is experimental but *ok*.
* I have started to write a simple backend (simpler but less efficient than Gorgonia without automatic differentiation). This has helped me to setup a more usable API for the interfaces.  `OperationApplyer` and `Operation` should be replaced by something more elegant:

```go
// OperationCarrier is any graph that can apply operations on its nodes
// regarding the structure of the graph
type OperationCarrier interface {
        NewOp(s string) (Op, error)
        Apply(operation Op, n graph.Node) error
}

// Op represents an operation to be applied on a node N of the graph G.
type Op interface {
        // Make the operation regarding the node's shape and its surrounding
        // elements inside the graph.
        // For example, the Make operation of the Add operator can take care of broadcasting.
        // please take care that within this method you can use shape, type and dimensions of
        // the children nodes, but (most of the time) not their actual values.
        //
        // WARNING: this is not the actual execution of the operation that should be performed
        // with whatever method at runtime while walking the graph. The order of execution
        // of the operators is not guaranteed at the build phase.
        Make(g graph.WeightedDirected, n graph.Node) error
}
```

I need to adapt Gorgonia to use this new API.

**Comment by owulveryck** [comment]
A lot of work has been done on onnx-go.
* The API is nearly stable
* onnx-go presents a top-level structure called [`Model`](https://godoc.org/github.com/owulveryck/onnx-go#Model)
  * The `Model` wraps a [`Backend`](https://godoc.org/github.com/owulveryck/onnx-go#Backend)  which is a receiver for the computation graph. 
  * The `Model` fulfills the [`encoding.BinaryUnmarshaler`](https://golang.org/pkg/encoding/#BinaryUnmarshaler)
* a `ComputationBackend` is an interface that adds the `Run()` method to a `Backend`. This is mandatory for being able to run backend that does symbolic computation (such as Gorgonia)
  * ComputationBackend can be tested thanks to a [`testbackend`](https://godoc.org/github.com/owulveryck/onnx-go/backend/testbackend) package and the `onnx` test files.
  * Gorgonnx is a special package that fulfills the `ComputationBackend` interface with Gorgonia.

**Comment by owulveryck** [comment]
I am closing this issue because the project is now stable enough
